src/math/ring_signature.cpp

```diff
@@ -0,0 +1,363 @@
+        // Is this public key in our secret keys map?
+        if (secret_keys.find(key) != secret_keys.end())
+            known_ring.push_back(key);
+    }
+    return known_ring;
+}
+
+// For all rings, make a list of known public keys corresponding to each ring.
+const key_rings partition_keys_into_rings(
+    const secret_keys_map& secret_keys,
+    const key_rings& rings)
+{
+    key_rings known_keys;
+    // For each ring in our rings...
+    for (const auto& ring: rings)
+    {
+        // ... find known keys corresponding to our private keys
+        const auto known_ring = known_keys_in_ring(secret_keys, ring);
+        // And add this list to our rings of known public keys.
+        known_keys.push_back(known_ring);
+    }
+    return known_keys;
+}
+
+// Check that all rings are not empty
+bool all_rings_nonzero_size(const key_rings& known_keys)
+{
+    for (const auto& ring: known_keys)
+        if (ring.empty())
+            return false;
+    return true;
+}
+
+typedef std::vector<size_t> index_list;
+
+// Make a list of indexes of where our known key occurs in each ring of
+// public keys. That is given a ring of {A, B, C} where we know the
+// private key of B, it will return 1 (the index in the ring).
+// This function computes this for all rings.
+index_list search_key_indexes(
+    const key_rings& rings,
+    const key_rings& known_keys_by_ring)
+{
+    index_list known_key_indexes;
+    // Iterate all our rings.
+    for (size_t i = 0; i < rings.size(); ++i)
+    {
+        BITCOIN_ASSERT(i < known_keys_by_ring.size());
+        const auto& ring = rings[i];
+        const auto& known = known_keys_by_ring[i];
+        // Find known key in this current ring.
+        const auto it = std::find(ring.begin(), ring.end(), known.back());
+        BITCOIN_ASSERT(it != ring.end());
+        // Calculate and store the index
+        size_t index = it - ring.begin();
+        known_key_indexes.push_back(index);
+    }
+    return known_key_indexes;
+}
+
+// Returns message || flatten(rings)
+data_chunk concatenate(const data_slice message, const key_rings& rings)
+{
+    data_chunk result(message.begin(), message.end());
+    for (const auto& ring: rings)
+        for (const auto& key: ring)
+            extend_data(result, key);
+    return result;
+}
+
+// Generate random k and s values.
+// There is a single k for each ring.
+// Each key in each ring, has a corresponding s value.
+// That is, each ring[i][j] should have a corresponding s[i][j]
+void random_k_and_s_values(
+    secret_list& k, ring_signature::s_values_type& s,
+    const key_rings& rings, const data_slice seed)
+{
+    // This value increments inside this function with every call
+    // to random_scalar() lambda.
+    uint32_t increment_randomizer = 0;
+    // Use a hd_private wallet as a random generator for ec_secret values.
+    auto random_scalar = [&seed, &increment_randomizer]()
+    {
+        static const wallet::hd_private generate_secret(to_chunk(seed));
```

Comment:
I'd prefer to not have a dependency on `bc::wallet` here. It might be worth at some point factoring out a seed->ec_secret function and placing into elliptic_curve.cpp.

Reply:
Yes, it's a bit hackish and not nice

Reply:
Note also that key generation can fail so there error must be handled (more info in bip32 and libsecp256k1).

Reply:
Maybe I'll just accept these parameters and for now let the user randomly generate them. Later we can add a wrapper which accepts a random seed.

Reply:
That just puts the complexity on the user. The library logically only needs the seed, so publicly exposing its internals to avoid that is suboptimal from an interface standpoint. BIP32 requires that you handle the potential invalidity of a generated key by skipping to the next value. That can be done here internally, although it does present a halting problem. So it’s best just to return a failure in the case where any one fails just fail the function. The chance of it is very low, but needs to be handled.